src/main/host/timer.rs

```diff
@@ -0,0 +1,237 @@
+use std::sync::{Arc, Weak};
+
+use atomic_refcell::AtomicRefCell;
+use log::trace;
+
+use crate::core::support::emulated_time::EmulatedTime;
+use crate::core::support::simulation_time::SimulationTime;
+use crate::core::work::task::TaskRef;
+use crate::core::worker::Worker;
+use crate::utility::{Magic, ObjectCounter};
+
+use super::host::Host;
+
+pub struct Timer {
+    magic: Magic<2537832161>,
+    _counter: ObjectCounter,
+    // Internals in an Arc so that we can schedule tasks that refer back to it.
+    // This is the only persistent strong reference - callbacks use a Weak
+    // reference.  i.e. dropping the outer object will drop this field as well;
+    // scheduled callbacks with weak references that can't be upgraded become
+    // no-ops.
+    internal: Arc<AtomicRefCell<TimerInternal>>,
+}
+
+struct TimerInternal {
+    next_expire_time: Option<EmulatedTime>,
+    expire_interval: SimulationTime,
+    expiration_count: u64,
+    next_expire_id: u64,
+    min_valid_expire_id: u64,
+    on_expire: Box<dyn Fn(&mut Host) + Send + Sync>,
+}
+
+impl Timer {
+    /// Create a new Timer that synchronously executes `on_expire` on
+    /// expiration.  `on_expire` will cause a panic if it calls mutable methods
+    /// of the enclosing Timer.  If it may need to call mutable methods of the
+    /// Timer, it should push a new task to the scheduler to do so.
+    pub fn new<F: 'static + Fn(&mut Host) + Send + Sync>(on_expire: F) -> Self {
+        Self {
+            magic: Magic::new(),
+            _counter: ObjectCounter::new("Timer"),
+            internal: Arc::new(AtomicRefCell::new(TimerInternal {
+                next_expire_time: None,
+                expire_interval: SimulationTime::ZERO,
+                expiration_count: 0,
+                next_expire_id: 0,
+                min_valid_expire_id: 0,
+                on_expire: Box::new(on_expire),
+            })),
+        }
+    }
+
+    pub fn expiration_count(&self) -> u64 {
+        self.magic.debug_check();
+        self.internal.borrow().expiration_count
+    }
+
+    pub fn expire_interval(&self) -> SimulationTime {
+        self.magic.debug_check();
+        self.internal.borrow().expire_interval
+    }
+
+    pub fn consume_expiration_count(&mut self) -> u64 {
+        self.magic.debug_check();
+        let mut internal = self.internal.borrow_mut();
+        let e = internal.expiration_count;
+        internal.expiration_count = 0;
+        e
+    }
+
+    pub fn remaining_time(&self) -> SimulationTime {
+        self.magic.debug_check();
+        let t = if let Some(t) = self.internal.borrow().next_expire_time {
+            t
+        } else {
+            return SimulationTime::ZERO;
```

Comment:
Function could use comments explaining that the remaining time is 0 for a timer with no expire time. Or should this return an `Option` instead?

Reply:
Done. Added comments and used an Option here. Pushed the "zero" special handling into the C interface.